pfcpiface/parser.go

```diff
@@ -0,0 +1,257 @@
+	var eNBTeid uint32
+	eNBIP := uint32(0)
+	n6IP4 := uint32(0)
+	tunnelType := uint8(0)
+	ies2, err := ie1.ForwardingParameters()
+	if err != nil {
+		log.Println("Unable to find ForwardingParameters!")
+	} else {
+		for _, ie2 := range ies2 {
+			switch ie2.Type {
+			case ie.OuterHeaderCreation:
+				outerheadercreationfields, err := ie2.OuterHeaderCreation()
+				if err != nil {
+					log.Println("Unable to parse OuterHeaderCreationFields!")
+					continue
+				}
+				eNBTeid = outerheadercreationfields.TEID
+				eNBIP = ip2int(outerheadercreationfields.IPv4Address)
+				n6IP4 = ip2int(n6IP)
+				tunnelType = uint8(1)
+			}
+		}
+	}
+	far := far{
+		farID:       uint8(farID),       // farID currently being truncated to uint8 <--- FIXIT/TODO/XXX
+		fseID:       uint32(fseid.SEID), // fseID currently being truncated to uint32 <--- FIXIT/TODO/XXX
+		action:      farForwardU,
+		tunnelType:  tunnelType,
+		s1uIP:       n6IP4,
+		eNBIP:       eNBIP,
+		eNBTeid:     eNBTeid,
+		UDPGTPUPort: udpGTPUPort,
+	}
+
+	return &far
+}
+
+func parseUpdateFAR(ie1 *ie.IE, fseid uint64, n3IP net.IP) (farI *far) {
+	/* check for updatefar, and fetch FAR ID */
+	farID, err := ie1.FARID()
+	if err != nil {
+		log.Println("Unable to find updateFAR's FAR ID!")
+		return nil
+	}
+	var eNBTeid uint32
+	eNBIP := uint32(0)
+	n6IP4 := uint32(0)
+	tunnelType := uint8(0)
+	/* Read UpdateFAR from payload */
+	ies2, err := ie1.UpdateForwardingParameters()
+	if err != nil {
+		log.Println("Unable to find UpdateForwardingParameters!")
+	} else {
+		for _, ie2 := range ies2 {
+			switch ie2.Type {
+			case ie.OuterHeaderCreation:
+				outerheadercreationfields, err := ie2.OuterHeaderCreation()
+				if err != nil {
+					log.Println("Unable to parse OuterHeaderCreationFields!")
+				} else {
+					eNBTeid = outerheadercreationfields.TEID
+					eNBIP = ip2int(outerheadercreationfields.IPv4Address)
+					n6IP4 = ip2int(n3IP)
+					tunnelType = uint8(1)
+				}
+			case ie.DestinationInterface:
+				// Do nothing for the time being
+			}
+		}
+	}
+	far := far{
+		farID:       uint8(farID),  // farID currently being truncated to uint8 <--- FIXIT/TODO/XXX
+		fseID:       uint32(fseid), // fseID currently being truncated to uint32 <--- FIXIT/TODO/XXX
+		action:      farForwardD,
+		tunnelType:  tunnelType,
+		s1uIP:       n6IP4,
+		eNBIP:       eNBIP,
+		eNBTeid:     eNBTeid,
+		UDPGTPUPort: udpGTPUPort,
+	}
+
+	return &far
+}
+
+func parsePDRFromPFCPSessEstReqPayload(upf *upf, sereq *message.SessionEstablishmentRequest, fseid *ie.FSEIDFields) (pdrs []pdr, fars []far) {
+
+	pdrList := make([]pdr, 0, MaxItems)
+	farList := make([]far, 0, MaxItems)
+
+	/* read PDR(s) */
+	ies1, err := ie.ParseMultiIEs(sereq.Payload)
+	if err != nil {
+		log.Println("Failed to parse sereq for IEs!")
+		return pdrList, farList
+	}
+	/*
+	 * Iteratively go through all IEs. You can't use ie.CreatePDR or ie.CreateFAR since a single
+	 * message can carry multiple CreatePDR & CreateFAR messages.
+	 */
```

Comment:
```suggestion
	 // Iteratively go through all IEs. You can't use ie.CreatePDR or ie.CreateFAR since a single
	 // message can carry multiple CreatePDR & CreateFAR messages.
```